src/util/Utils.js

```diff
@@ -720,7 +720,7 @@ x3dom.Utils.isUnsignedType = function (str)
 *****************************************************************************/
 x3dom.Utils.checkDirtyLighting = function(viewarea)
 {
-	return (viewarea.getLights().length + viewarea._scene.getNavigationInfo()._vf.headlight);
+	return (viewarea.getLights().concat(viewarea._scene.getNavigationInfo()._vf.headlight));
```

Comment:
With this check all the shaders are regenerated every frame because it will compare two pointers to the same array.

Simultaneous adding and removing needs no shader regeneration.
The PhysicalEnvironmentLight is handled diffrent and needs an additional check.

Reply:
hm, right. There is the case that the env. light is removed and the headlight is added. This will result in no change in the count, and no regeneration although it is then needed, no ?

Reply:
I would add following function at line 736 in Utils.js

```
/*****************************************************************************
 * Checks for PhysicalEnvironmentLight change
 *****************************************************************************/
x3dom.Utils.checkDirtyPhysicalEnvironmentLight = function(viewarea, shaderProperties)
{
    return (!!shaderProperties.PHYSICALENVLIGHT == viewarea.hasPhysicalEnvironmentLight());
};
```

and change the getShaderProperties function in X3DShapeNode.js to:

```
getShaderProperties: function(viewarea)
{
    if (this._shaderProperties == null ||
        this._dirty.shader == true ||
        x3dom.Utils.checkDirtyEnvironment(viewarea, this._shaderProperties) ||
        x3dom.Utils.checkDirtyPhysicalEnvironmentLight (viewarea, this._shaderProperties) ||
        (this._webgl !== undefined &&
         this._webgl.dirtyLighting != x3dom.Utils.checkDirtyLighting(viewarea) ))
    {
        this._shaderProperties = x3dom.Utils.generateProperties(viewarea, this);
        this._dirty.shader = false;
        if (this._webgl !== undefined)
        {
             this._webgl.dirtyLighting = x3dom.Utils.checkDirtyLighting(viewarea);
        }
    }
    return this._shaderProperties;
},
```

Reply:
agreed and tested